src/hdmf_zarr/backend.py

```diff
@@ -1099,11 +1099,16 @@ def __read_dataset(self, zarr_obj, name):
         if ret is not None:
             return ret
 
-        if 'zarr_dtype' not in zarr_obj.attrs:
-            raise ValueError("Dataset missing zarr_dtype: " + str(name) + "   " + str(zarr_obj))
+        if hasattr(zarr_obj, 'dtype'):
+            zarr_dtype = zarr_obj.dtype
+        else:
+            if 'zarr_dtype' in zarr_obj.attrs:
+                zarr_dtype = zarr_obj.attrs['zarr_dtype']
+            else:
+                raise ValueError("Dataset missing zarr_dtype: " + str(name) + "   " + str(zarr_obj))
```

Comment:
I would suggest having the case where we use `zarr_dtype` attribute first and using the `zarr_obj.dtype` as a fallback, instead of the other way around. Also, I would suggest raising a warning when the dtype is inferred, just in case this should lead to any downstream errors to ease debugging and to make sure the user is still notified that the file has a problem.

```suggestion
        if 'zarr_dtype' not in zarr_obj.attrs:
            zarr_dtype = zarr_obj.attrs['zarr_dtype']   
        elif hasattr(zarr_obj, 'dtype'):   # Fallback for invalid files that are mssing zarr_type
            zarr_dtype = zarr_obj.dtype
            warnings.warn("Inferred dtype from zarr type. Dataset missing zarr_dtype: " + str(name) + "   " + str(zarr_obj)) 
        else:     
            raise ValueError("Dataset missing zarr_dtype: " + str(name) + "   " + str(zarr_obj))
```

Reply:
done